dcm4che-core/src/main/java/org/dcm4che3/io/DicomInputStream.java

```diff
@@ -484,29 +491,40 @@ public Attributes readCommand() throws IOException {
     }
 
     public void readAllAttributes(Attributes attrs) throws IOException {
-        readAttributes(attrs, -1, o -> false);
+        readAttributes(attrs, UNDEFINED_LENGTH, o -> false);
     }
 
     public Attributes readDataset() throws IOException {
-        return readDataset(-1, o -> false);
+        return readDataset(UNDEFINED_LENGTH, o -> false);
+    }
+
+    public Attributes readDataset(long len) throws IOException {
+        return readDataset(len, o -> false);
     }
 
     public Attributes readDatasetUntilPixelData() throws IOException {
-        return readDataset(-1, o -> o.tag == Tag.PixelData);
+        return readDataset(UNDEFINED_LENGTH, o -> o.tag == Tag.PixelData);
     }
 
     @Deprecated
-    public Attributes readDataset(int len, int stopTag) throws IOException {
+    public Attributes readDataset(long len, int stopTag) throws IOException {
         return readDataset(len, tagEqualOrGreater(stopTag));
     }
 
-    public Attributes readDataset(int len, Predicate<DicomInputStream> stopPredicate) throws IOException {
+    public Attributes readDataset(long len, Predicate<DicomInputStream> stopPredicate) throws IOException {
         handler.startDataset(this);
+        readDatasetEndPos = (len == UNDEFINED_LENGTH) ? UNDEFINED_LENGTH : pos + len;
+        long posBeforeReadingFMI = pos;
         readFileMetaInformation();
+        if(len != UNDEFINED_LENGTH) {
+            len -= (pos - posBeforeReadingFMI);
```

Comment:
I'm correcting the `len` here now, in the case fmi was not read yet.

But I am a bit unsure about the whole thing. (See my long previous comment.)

Open points:
* What should we do about the problem with the bytes already read in the constructor. Is it really the user of the API that has to accommodate for that?
* Extend DicomInputStreamTest to cover also reading with a `len` given for different test-data cases.
* Fix the deflated case.